lib/iaas/GcpClient.js

```diff
@@ -0,0 +1,200 @@
+      })
+      .catchThrow(BaseCloudClient.providerErrorTypes.NotFound,
+        new NotFound(`Object '${file}' not found while deleting in container ${container}`));
+  }
+
+  download(options) {
+    return utils.streamToPromise(this.storageClient
+        .bucket(options.container)
+        .file(options.remote)
+        .createReadStream())
+      .catchThrow(BaseCloudClient.providerErrorTypes.NotFound, new NotFound(`Object '${options.remote}' not found`));
+  }
+
+  upload(options, buffer) {
+    return new Promise((resolve, reject) => {
+      function cleanup() {
+        uploadedStream.removeListener('finish', onfinish);
+        uploadedStream.removeListener('error', onerror);
+      }
+
+      function onerror(err) {
+        cleanup();
+        reject(err);
+      }
+
+      function onfinish() {
+        cleanup();
+        resolve(JSON.parse(buffer));
+      }
+
+      let uploadedStream = this.storageClient
+        .bucket(options.container)
+        .file(options.remote)
+        .createWriteStream();
+
+      uploadedStream.on('error', onerror);
+      uploadedStream.on('finish', onfinish);
+      uploadedStream.end(buffer);
+    });
+  }
+
+  uploadJson(container, file, data) {
+    if (arguments.length < 3) {
+      data = file;
+      file = container;
+      container = this.containerName;
+    }
+    return this
+      .upload({
+        container: container,
+        remote: file
+      }, new Buffer(JSON.stringify(data, null, 2), 'utf8'));
+  }
+
+  downloadJson(container, file) {
+    if (arguments.length < 2) {
+      file = container;
+      container = this.containerName;
+    }
+    return this
+      .download({
+        container: container,
+        remote: file
+      })
+      .then((data) => JSON.parse(data))
+      .catchThrow(SyntaxError, new NotFound(`Object '${file}' not found`));
+  }
+
+  deleteSnapshot(snapshotName) {
+    return Promise
+      .try(() => {
+        return this.computeClient
+          .snapshot(snapshotName)
+          .delete()
+          .then(retval => logger.info(`Deleted snapshot ${snapshotName}`, retval));
+      })
+      .catch(BaseCloudClient.providerErrorTypes.Unauthorized, err => {
+        logger.error(err.message);
+        throw new Unauthorized(`Authorization at google compute failed while deleting snapshot ${snapshotName}`);
+      })
+      .catch(BaseCloudClient.providerErrorTypes.Forbidden, err => {
+        logger.error(err.message);
+        throw new Forbidden(`Authentication at google compute failed while deleting snapshot ${snapshotName}`);
+      })
+      .catchThrow(BaseCloudClient.providerErrorTypes.NotFound,
+        new NotFound(`Snapshot ${snapshotName} not found while deleting`));
+  }
+
+  static validateParams(options) {
+    if (!options) {
+      throw new Error('GcpClient can not be instantiated as backup provider config not found');
+    } else {
+      if (!options.projectId || !options.credentials) {
```

Comment:
Dont you also want to check the validity of `credentials` object?

Reply:
There are multiple fields in the credentials object like type, client_email, client_id etc. and this is directly imported from a key file generated via GCP console / cli . Hence I do not think there is a need to validate the credentials object further.

Reply:
As discussed, I will also check for individual parameters in validate.

Reply:
ok